File: src/main/java/com/example/diem_danh/service/ChatService.java
package com.example.diem_danh.service;

import com.example.diem_danh.dto.request.CreateGroupRequest;
import com.example.diem_danh.dto.request.SendMessageRequest;
import com.example.diem_danh.dto.response.ConversationResponse;
import com.example.diem_danh.dto.response.MessageResponse;
import com.example.diem_danh.dto.response.UserResponse;

import java.util.List;

public interface ChatService {

    // ── Conversation ──────────────────────────────────────────────────────────
    ConversationResponse getOrCreatePrivateConversation(String currentUserId, String targetUserId);

    /** Tạo GROUP — chỉ TEACHER/ADMIN */
    ConversationResponse createGroup(String currentUserId, String currentUserRole, CreateGroupRequest req);

    /**
     * Tạo CLASS conversation tự động khi tạo lớp học.
     * Gọi nội bộ từ ClassRoomService.
     */
    ConversationResponse createClassConversation(String classId, String className,
                                                  String teacherUserId, List<String> studentUserIds);

    /**
     * Thêm 1 sinh viên vào CLASS conversation của lớp khi enroll.
     * Gọi nội bộ từ ClassRoomService.
     */
    void addMemberToClassConversation(String classId, String studentUserId);

    void deleteGroup(String currentUserId, String conversationId);
    void leaveGroup(String currentUserId, String conversationId);
    void addMember(String currentUserId, String conversationId, String targetUserId);
    void addMembers(String currentUserId, String conversationId, List<String> targetUserIds);
    void removeMember(String currentUserId, String conversationId, String targetUserId);
    void promoteAdmin(String currentUserId, String conversationId, String targetUserId);
    void demoteAdmin(String currentUserId, String conversationId, String targetUserId);
    void renameGroup(String currentUserId, String conversationId, String newName);

    List<ConversationResponse> getMyConversations(String userId, String userRole);
    ConversationResponse getConversation(String conversationId);

    // ── Members info ──────────────────────────────────────────────────────────
    List<UserResponse> getConversationMembers(String conversationId);
    List<UserResponse> searchUsersForChat(String keyword);

    // ── Messages ──────────────────────────────────────────────────────────────
    MessageResponse sendMessage(String senderId, SendMessageRequest req);
    MessageResponse editMessage(String callerId, String messageId, String newContent);
    MessageResponse deleteMessage(String callerId, String messageId);

    /** Ghim — chỉ TEACHER hoặc ADMIN (system role) */
    MessageResponse pinMessage(String callerId, String callerRole, String messageId);

    /** Bỏ ghim — chỉ TEACHER hoặc ADMIN (system role) */
    MessageResponse unpinMessage(String callerId, String callerRole, String messageId);

    List<MessageResponse> getMessages(String conversationId, int page, int size);
    List<MessageResponse> searchMessages(String conversationId, String keyword);
    List<MessageResponse> getPinnedMessages(String conversationId);

    // Backward-compat alias
    default ConversationResponse getOrCreateDirectConversation(String currentUserId, String targetUserId) {
        return getOrCreatePrivateConversation(currentUserId, targetUserId);
    }
}

File: src/services/chat.service.ts
import api from './api'
import type {
  Conversation,
  Message,
  SendMessagePayload,
  CreateGroupPayload,
  ChatUser,
} from '../models/chat.model'

const BASE = '/chat'

export const chatService = {
  // ── Conversations ─────────────────────────────────────────────────────────
  getMyConversations: () =>
    api.get<{ data: Conversation[] }>(`${BASE}/conversations`).then(r => r.data.data),

  openDirect: (targetUserId: string) =>
    api.post<{ data: Conversation }>(`${BASE}/conversations/direct/${targetUserId}`).then(r => r.data.data),

  /** Tạo nhóm GROUP — chỉ TEACHER/ADMIN có quyền, backend enforce */
  createGroup: (payload: CreateGroupPayload) =>
    api.post<{ data: Conversation }>(`${BASE}/conversations/group`, payload).then(r => r.data.data),

  getConversation: (conversationId: string) =>
    api.get<{ data: Conversation }>(`${BASE}/conversations/${conversationId}`).then(r => r.data.data),

  deleteGroup: (conversationId: string) =>
    api.delete(`${BASE}/conversations/${conversationId}`),

  leaveGroup: (conversationId: string) =>
    api.post(`${BASE}/conversations/${conversationId}/leave`),

  /** Đổi tên nhóm — chỉ admin nhóm (group role) */
  renameGroup: (conversationId: string, name: string) =>
    api.patch(`${BASE}/conversations/${conversationId}/name`, { name }),

  // ── Member management ──────────────────────────────────────────────────────
  getMembers: (conversationId: string): Promise<ChatUser[]> =>
    api.get<{ data: ChatUser[] }>(`${BASE}/conversations/${conversationId}/members`).then(r => r.data.data),

  addMember: (conversationId: string, targetUserId: string) =>
    api.post(`${BASE}/conversations/${conversationId}/members/${targetUserId}`),

  addMembers: (conversationId: string, userIds: string[]) =>
    api.post(`${BASE}/conversations/${conversationId}/members`, { userIds }),

  removeMember: (conversationId: string, targetUserId: string) =>
    api.delete(`${BASE}/conversations/${conversationId}/members/${targetUserId}`),

  promoteAdmin: (conversationId: string, targetUserId: string) =>
    api.post(`${BASE}/conversations/${conversationId}/members/${targetUserId}/promote`),

  demoteAdmin: (conversationId: string, targetUserId: string) =>
    api.post(`${BASE}/conversations/${conversationId}/members/${targetUserId}/demote`),

  // ── User search for chat picker ────────────────────────────────────────────
  searchUsers: (keyword?: string): Promise<ChatUser[]> =>
    api.get<{ data: ChatUser[] }>(`${BASE}/users`, { params: { keyword } }).then(r => r.data.data),

  // ── Messages ──────────────────────────────────────────────────────────────
  getMessages: (conversationId: string, page = 0, size = 30) =>
    api
      .get<{ data: Message[] }>(`${BASE}/conversations/${conversationId}/messages`, {
        params: { page, size },
      })
      .then(r => r.data.data),

  sendMessage: (payload: SendMessagePayload) =>
    api.post<{ data: Message }>(`${BASE}/messages`, payload).then(r => r.data.data),

  editMessage: (messageId: string, content: string) =>
    api.put<{ data: Message }>(`${BASE}/messages/${messageId}`, { content }).then(r => r.data.data),

  deleteMessage: (messageId: string) =>
    api.delete<{ data: Message }>(`${BASE}/messages/${messageId}`).then(r => r.data.data),

  /**
   * Ghim tin nhắn — chỉ TEACHER hoặc ADMIN (system role).
   * Backend sẽ kiểm tra lại và trả về 403 nếu không đủ quyền.
   * POST /api/chat/messages/{id}/pin
   */
  pinMessage: (messageId: string) =>
    api.post<{ data: Message }>(`${BASE}/messages/${messageId}/pin`).then(r => r.data.data),

  /**
   * Bỏ ghim tin nhắn — chỉ TEACHER hoặc ADMIN (system role).
   * POST /api/chat/messages/{id}/unpin
   */
  unpinMessage: (messageId: string) =>
    api.post<{ data: Message }>(`${BASE}/messages/${messageId}/unpin`).then(r => r.data.data),

  searchMessages: (conversationId: string, keyword: string) =>
    api
      .get<{ data: Message[] }>(`${BASE}/conversations/${conversationId}/messages/search`, {
        params: { keyword },
      })
      .then(r => r.data.data),

  /**
   * Lấy danh sách tin nhắn đã ghim.
   * Sắp xếp theo pinnedAt giảm dần (mới nhất trước).
   * GET /api/chat/conversations/{id}/pinned-messages
   */
  getPinnedMessages: (conversationId: string) =>
    api
      .get<{ data: Message[] }>(`${BASE}/conversations/${conversationId}/pinned-messages`)
      .then(r => r.data.data),

  uploadFile: async (file: File): Promise<{ fileUrl: string; fileName: string }> => {
    const form = new FormData()
    form.append('file', file)
    const r = await api.post<{ data: { fileUrl: string; fileName: string } }>(`${BASE}/upload`, form, {
      headers: { 'Content-Type': 'multipart/form-data' },
    })
    return r.data.data
  },
}
